### ISSUE TITLE
Prefer typescript solutions, e.g discriminated unions, over console warnings for consumers

### ISSUE THREAD

**Issue by OliBridgman** [cultureamp/kaizen-design-system#1834]
I'm not sure of the size of this issue, but I'll focus on Icons as an example.

Icons of type `img` require an accessible title. Icons of type `presentation` do not. To nudge consumers, we currently warn when an icon of type `img` is used _without_ a title.

```
if (isMeaningfulImg && !title) {
  warn(`
    Icon with role "img" missing a title attribute.

    Assistive technologies that enable vision-impaired users to read web pages
    can treat images (including icons) as either decorative or meaningful. Only
    images with a role of "img" (meaningful) will be read aloud to the user.
    These images must therefore have a title attribute to provide the text that
    will be read aloud.

    Either add the missing title prop, or set this icon's role to
    "presentation" to indicate it is not meaningful.
  `)
}
```

Often, this gets ignored. Warnings can end up in production:
<img width="891" alt="Screen Shot 2021-08-11 at 11 48 48 AM" src="https://user-images.githubusercontent.com/1510005/128949151-082a4f1e-5d93-42e4-8b54-9b2d21ac1ea1.png">

They can also spam the developer console locally and they pollute the console during testing.

The warnings don't _really_ seem to be doing their job (although we only see when it doesn't work so maybe that statement is a lie) and they cause frustrations for others who share the codebase.

As an alternative, we could enforce the relationship these props have with discriminated union types (or something even fancier?).

Here is a quick example (not too much thought put into this, not necessarily the solution):

```
// meaningful, title should be read aloud to users who can't see it
type ImageIconProps = { role: "img"; title: string }
// decorative, should be silent to users who can't see it
type PresentationIconProps = { role: "presentation"; title?: never }

type IconProps = React.FunctionComponent<
  {
    icon: React.SVGAttributes<SVGSymbolElement>
    inheritSize?: boolean
    desc?: string
  } & (ImageIconProps | PresentationIconProps)
>
```

With something like this, the consumer "can't" (obviously they can do anything cos it's JustJavascript™) render an inaccessible icon.

https://user-images.githubusercontent.com/1510005/128959927-023266ad-ea80-40ab-880d-3a7c37d085be.mp4

The example is specific to icons and my current annoyance with the console logs, but there are possibly more places this could be useful where props exist in conflict with each other with implicit seperate use cases that we could make more explicit with typescript. Whether or not this issue should be specific to Icons or be about the idea in general, I dunno.

**Comment by jasononeil** [comment]
One thing to be careful of is that in many of these warnings, we're checking for both null, undefined, and an empty string. 

I explored a way to pull this off in a comment over in a different issue: https://github.com/cultureamp/kaizen-design-system/issues/1530#issuecomment-881141049
